### ISSUE TITLE
Statefulness - Contiguous Data Across Batches and Backpropagation

### ISSUE THREAD

**Issue by Null-Quark** [MUST-Deep-Learning/KnowIt#131]
**Problem:** Currently, KnowIt's Trainer and Data module is unable to handle contiguous data across batches as well as implementing parameter updates across batches (instead of per batch).

In the current setting, batches containing shuffled chunks of data are used to perform backpropagation. This assumes that:

1.  The sample in the batch has a sufficient window size to allow the model to learn important patterns.
2. The time step window for each sample inside a batch is sufficiently small so that the data can fit into memory.

However, there exists use-cases where users need very large contiguous time step windows; consequently, batches may not be able to fit into memory. To overcome this, a single large batch is split into smaller batches that are contiguous in time. During learning, gradients graphs need to be retained across the smaller batches. This is in particular important for RNNs which have hidden states that retains important information about the past.

**Task:** To implement such a functionality: 

1. The data module would need to accommodate batches where data remains contiguous between batches (data is assumed to be a single large instance; in other words, there are no gaps in the data). The training data should not be shuffled.
2. The trainer needs modification: RNN-type models such as LSTMs would need to preserve the hidden states across batches to allow truncated backprop to be performed (this can be done with the Pytorch setting `retain_graph=True.` Currently, the LSTM resets the hidden/cell states for each new batch, which is standard practice.
